### ISSUE TITLE
BFGMiner exhausts memory with threads when network is down or failing

### ISSUE THREAD

**Issue by jddebug** [luke-jr/bfgminer#11]
On Windows, this manifests as a crash often in curl_maprintf.
On Linux, pthread_create fails and BFGMiner quits.

Original report from @jddebug:

bfgminer crashes after several hours on win xp with a mini rig. Only happens on the 19 card mini so far. Will add more as I learn more.

_Possibly also affecting: eryngi and Bunnyh_

**Comment by jddebug** [comment]
This was the crash report.

```
bfgminer.exe caused an Access Violation at location 70817995 in module libcurl-4.dll Reading from location 0001312d.

Registers:
eax=003e75a8 ebx=003e75a8 ecx=77c2c2e3 edx=00000000 esi=0001312d edi=003e75a8
eip=70817995 esp=268ff248 ebp=268ff384 iopl=0         nv up ei pl nz na po nc
cs=001b  ss=0023  ds=0023  es=0023  fs=003b  gs=0000             efl=00000206

Call stack:
70817995  libcurl-4.dll:70817995  curl_maprintf
7C90F668  ntdll.dll:7C90F668  RtlNtStatusToDosError
565308EC
```

The api thinks the miner is still running and reports good hash rates but it has an error screen and is not actually running.

**Comment by luke-jr** [comment]
To test:
- BFGMiner 2.4.4
- Copy libcurl-4.dll from CGMiner

**Comment by jddebug** [comment]
Here are some pictures of the screen when bfgminer crashed numerous times today. Both are windows 7 machines.

Really hope we can get this fixed.

Thanks

**Comment by jddebug** [comment]
Another couple screenshots from this morning when I got up. Any thoughts at all on how to fix?

**Comment by luke-jr** [comment]
Where are the screenshots? :/

**Comment by jddebug** [comment]
bfgminer.exe caused an Access Violation at location 70817995 in module libcurl-4.dll Reading from location 0002dc6c.

Registers:
eax=003e7730 ebx=003e7730 ecx=77c2c2e3 edx=00000000 esi=0002dc6c edi=003e7730
eip=70817995 esp=2147f248 ebp=2147f384 iopl=0         nv up ei pl nz na po nc
cs=001b  ss=0023  ds=0023  es=0023  fs=003b  gs=0000             efl=00000206

Call stack:
70817995  libcurl-4.dll:70817995  curl_maprintf
7C90F668  ntdll.dll:7C90F668  RtlNtStatusToDosError
565308EC

**Comment by luke-jr** [comment]
Identified bugs:
- Number of get_work threads may grow unbounded in some cases. This is hard to reproduce, as the submissions usually shut us down first.
- Number of submission threads may grow unreasonably high during network outages, or potentially even ordinarily with enough hashrate.

**Comment by luke-jr** [comment]
Fixed in 06041dd and 57aa649
